Add IPFilter to restrict access to allowed list IPs.

felipebonezi opened the pull request:
# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [x] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Feature

Closes #10483 

## Purpose

Add IPFilter to restrict access to allowed list IPs.

## Background Context

I've created an EssentialFilter which checks all upcoming HTTP requests if the Remote Addres (IP) is allowed to access the requested route.

## References

Issue #10483

felipebonezi commented:
@mkurz can you help me?
I did used the `sbt scalafmtAll scalafmtSbt` like Travis is recommending.

felipebonezi commented:
> @mkurz can you help me?
> I did used the `sbt scalafmtAll scalafmtSbt` like Travis is recommending.

@mkurz sorry, I did! 🎉

mkurz commented on documentation/manual/working/commonGuide/filters/AllowedIPFilter.md:
````diff
@@ -0,0 +1,31 @@
+If you don't want to check the HTTP request remote address to any route path you can use `play.filters.ip.excludePaths`.
+
+```
+play.filters.ip.excludePaths = [ '/path1', '/path2' ]
````
Instead (or in addition?) of this `excludePaths` config I think it would be better if we make use of Play's route modifier tags. So you can define excluded paths directly in the routes, like:
```
+ noipcheck
GET /path1 MyController.foo
+ noipcheck
GET /path2 MyController.bar
```
I think this is safer than the config, because if you change the path in the routes file, you may forget to also change the path in the `excludePaths` config.
What do you think?

(More information about route modifier tags can be found here: https://www.playframework.com/documentation/2.8.x/Highlights26#Route-modifier-tags)

felipebonezi replied:
Done @mkurz ✅

mkurz replied:
You still have `excludePaths` documented, but it's not implemented anymore.

mkurz reviewed (changes requested):
Thanks for taking time and working on this!
I think it would be a good idea to also handle IPv6 adresses. What do you think?

felipebonezi commented:
@mkurz documentation fixed ✅

felipebonezi commented:
@mkurz what about now? Is everything ok?

mkurz commented:
@felipebonezi Yes, everything OK, I am just very busy right now. I think also @octonato and/or @ignasi35 should review too however.

felipebonezi commented:
@mkurz @octonato @ignasi35 hey guys, good afternoon.
I know that you're very busy but do you think this PR has any chance to goes with 2.9.x version? I really want to use this in my application! 👍🏻 

Thanks for your time, cya.

felipebonezi commented:
@mkurz hi, don't forget to review this PR. 😁

ruyconolly commented:
@mkurz is there any expectation when 2.9.0 will be released?

mkurz commented:
@felipebonezi @ruyconolly Sorry for the delay. So unfortunately I will be (again) quite busy with other stuff the next couple of weeks, probably til end of June (I will be around, but it could take a couple of days for responses). In the last steering meeting we targeted a first M1 or RC1 (depending on the state) in July. We will see if that holds. (It still also depends on other libraries like akka-http)
However, personally, I think about to put out a personal statement to let the community know what the situation is right now and be more transparent. Timewise my situation will fortunately finally get much much better starting with July and will only become better the months afterwards. Also we have enough money now from sponsorship which allows me to gradually start work full time on Play (still sponsoring needed, but I can use the time that the current opencollective income provides me to look for more). Like said I think about to communicate this clearly so everybody knows what's going on.

felipebonezi commented:
I squashed my commit into one and `Check / Code Style Docs / { "java": "8" }` is throwing that `AllowedIPFilter.md` hasn't a valid header but it has a valid one!

Is there anything to fix it?

https://github.com/playframework/playframework/pull/10536/files#diff-3ca5214a5275c47b22a5cfff9e0c81aa5803966bf8606ec295d11382646c98f7

BillyAutrey commented on documentation/manual/working/commonGuide/filters/AllowedIPFilter.md:
```diff
@@ -0,0 +1,33 @@
+<!--- Copyright (C) Lightbend Inc. <https://www.lightbend.com> -->
```
```suggestion
<!--- Copyright (C) Lightbend Inc. <https://www.lightbend.com> -->

```

BillyAutrey commented:
> I squashed my commit into one and `Check / Code Style Docs / { "java": "8" }` is throwing that `AllowedIPFilter.md` hasn't a valid header but it has a valid one!
> 
> Is there anything to fix it?
> 
> https://github.com/playframework/playframework/pull/10536/files#diff-3ca5214a5275c47b22a5cfff9e0c81aa5803966bf8606ec295d11382646c98f7

Normally I would suggest running `sbt headerCheckAll` from the `documentation` directory to validate.  But, I think your branch is a bit behind master, so some of the validations have changed.

In any case, I think that the issue is pretty simple here - it just expects a new line in between your header and title.

felipebonezi commented:
> > I squashed my commit into one and `Check / Code Style Docs / { "java": "8" }` is throwing that `AllowedIPFilter.md` hasn't a valid header but it has a valid one!
> > Is there anything to fix it?
> > https://github.com/playframework/playframework/pull/10536/files#diff-3ca5214a5275c47b22a5cfff9e0c81aa5803966bf8606ec295d11382646c98f7
> 
> Normally I would suggest running `sbt headerCheckAll` from the `documentation` directory to validate. But, I think your branch is a bit behind master, so some of the validations have changed.
> 
> In any case, I think that the issue is pretty simple here - it just expects a new line in between your header and title.

Thanks, worked.
I've rebased my branch from master as well.

felipebonezi commented:
@BillyAutrey I realize that the documentation project has a script called `addMarkdownCopyright`.
I executed this script but it doesn't add a new line between header and title... Do you think that we need to fix it?

mkurz commented:
@felipebonezi The `addMarkdownCopyright` script does not exist anymore, I removed it a couple of weeks ago. It is not present anymore in the main branch: https://github.com/playframework/playframework/tree/main/documentation
It was an old script, added long before we switched to sbt-header (I think even long before sbt-header came to live actually).
All you need to run ist `headerCheckAll` (or `headerCreateAll`).
Just make sure you are on latest master and everything is up to date.

felipebonezi commented:
> @felipebonezi The `addMarkdownCopyright` script does not exist anymore, I removed it a couple of weeks ago. It is not present anymore in the main branch: https://github.com/playframework/playframework/tree/main/documentation It was an old script, added long before we switched to sbt-header (I think even long before sbt-header came to live actually). All you need to run ist `headerCheckAll` (or `headerCreateAll`). Just make sure you are on latest master and everything is up to date.

Everything is done and ready for merge ✅

mkurz commented:
So I had another look at this pull request and there a some things that to change:

- First, you should put all the config keys you introduce into [this `reference.conf`](https://github.com/playframework/playframework/blob/main/web/play-filters-helpers/src/main/resources/reference.conf) file.

- You use 403 as response code, I think it would be good if we can also configure that, maybe someone wants to return 404 to "hide" an endpoint.

- You implemented an `allowList`. In Play, have the concept of black- and whitelists, which gives the user even more control. I think we should use the same for the ip filter: If you whitelist ips, only those ips can access a service. If the whitelist is empty and you have a blacklist, then all ips are allowed except the blacklisted ones.
We use that concept already in the csp, gzip, csrf and allowedhost filters.
See here the csrf config:
https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L79-L109
csp config:
https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L155-L162
allowed host filter:
https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L222-L231
gzip filter:
https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L274-L282
Just have a look at the gzip filter [how that is implemented](https://github.com/playframework/playframework/blob/5403d37e40da29ebf233c3a85ce1f17f780f0158/web/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala#L277-L294), or at any other of the above filter implementions: https://github.com/playframework/playframework/tree/main/web/play-filters-helpers/src/main/scala/play/filters

- Please update copyright headers, they changed now, see (maybe just rebased the pull request on current main)

Thank you very much!

felipebonezi commented:
> So I had another look at this pull request and there a some things that to change:
> 
> * First, you should put all the config keys you introduce into [this `reference.conf`](https://github.com/playframework/playframework/blob/main/web/play-filters-helpers/src/main/resources/reference.conf) file.
> * You use 403 as response code, I think it would be good if we can also configure that, maybe someone wants to return 404 to "hide" an endpoint.
> * You implemented an `allowList`. In Play, have the concept of black- and whitelists, which gives the user even more control. I think we should use the same for the ip filter: If you whitelist ips, only those ips can access a service. If the whitelist is empty and you have a blacklist, then all ips are allowed except the blacklisted ones.
>   We use that concept already in the csp, gzip, csrf and allowedhost filters.
>   See here the csrf config:
>   https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L79-L109
>   
>   csp config:
>   https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L155-L162
>   
>   allowed host filter:
>   https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L222-L231
>   
>   gzip filter:
>   https://github.com/playframework/playframework/blob/d7e969a66279aa49f4ff92ff7113f1b70b2b72b0/web/play-filters-helpers/src/main/resources/reference.conf#L274-L282
>   
>   Just have a look at the gzip filter [how that is implemented](https://github.com/playframework/playframework/blob/5403d37e40da29ebf233c3a85ce1f17f780f0158/web/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala#L277-L294), or at any other of the above filter implementions: https://github.com/playframework/playframework/tree/main/web/play-filters-helpers/src/main/scala/play/filters
> * Please update copyright headers, they changed now, see (maybe just rebased the pull request on current main)
> 
> Thank you very much!

Done, @mkurz ✅ 
Thanks for the feedback.

felipebonezi commented:
A new comment was made by @tsuyoshizawa privately.
https://github.com/felipebonezi/playframework/commit/ed7e19e1430afa601a8b65691cfb1df2635f15ff#r90546559

IMHO makes sense for HTML but for API I would recommend using only HTTP status code.
What do you think, @mkurz ?

tsuyoshizawa commented:
I was in the wrong place to comment. Thanks for adding the reference to my comment!

mkurz commented:
@felipebonezi I reviewed once more and pushed a couple of commits onto our commit, I hope you don't mind (you just need to run git fetch to get my changes). If you look at the commits one after the other, you hopefully understand what my intention was, also the commit message I added should clarify my intention.

@felipebonezi Can you please do one more thing? I saw that you added a `object IPKeys` which just holds the config keys. I think it's a better approach if we add a companion object below to the `class IPFilterConfig` with a `fromConfiguration` method. So we can easily call `IPFilterConfig.fromConfiguration(...)`. We do the same e.g. in the gzip filter and the allowed hosts filter, see:

* https://github.com/playframework/playframework/blob/172418b9acea4aaea8f4ad97b425cc980e3aa238/web/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala#L103-L108
* https://github.com/playframework/playframework/blob/172418b9acea4aaea8f4ad97b425cc980e3aa238/web/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala#L227-L230

After that I will take another fresh look. Thanks!

@tsuyoshizawa I made an http error go through the default http error handler now. So if the app is an api you probably defined the json error handler, so that should work the way intended: https://www.playframework.com/documentation/2.8.x/ScalaErrorHandling#Handling-errors-in-a-JSON-API

felipebonezi commented:
> @felipebonezi I reviewed once more and pushed a couple of commits onto our commit, I hope you don't mind (you just need to run git fetch to get my changes). If you look at the commits one after the other, you hopefully understand what my intention was, also the commit message I added should clarify my intention.
> 
> 
> 
> @felipebonezi Can you please do one more thing? I saw that you added a `object IPKeys` which just holds the config keys. I think it's a better approach if we add a companion object below to the `class IPFilterConfig` with a `fromConfiguration` method. So we can easily call `IPFilterConfig.fromConfiguration(...)`. We do the same e.g. in the gzip filter and the allowed hosts filter, see:
> 
> 
> 
> * https://github.com/playframework/playframework/blob/172418b9acea4aaea8f4ad97b425cc980e3aa238/web/play-filters-helpers/src/main/scala/play/filters/hosts/AllowedHostsFilter.scala#L103-L108
> 
> * https://github.com/playframework/playframework/blob/172418b9acea4aaea8f4ad97b425cc980e3aa238/web/play-filters-helpers/src/main/scala/play/filters/gzip/GzipFilter.scala#L227-L230
> 
> 
> 
> After that I will take another fresh look. Thanks!
> 
> 
> 
> @tsuyoshizawa I made an http error go through the default http error handler now. So if the app is an api you probably defined the json error handler, so that should work the way intended: https://www.playframework.com/documentation/2.8.x/ScalaErrorHandling#Handling-errors-in-a-JSON-API

I appreciate your review, thanks!
Everything is done as you asked for. ✅🙃

mkurz commented on web/play-filters-helpers/src/main/resources/reference.conf:
```diff
@@ -323,4 +324,30 @@ play.filters {
+      # If non empty, then requests will be checked if the route contains this modifier
+      # The black list is used only if the white list is empty
+      blackList = []
+    }
```
You can see in this `reference.conf` we already have such `routeModifiers.whiteList/blackList` configs for other filters. This allows the users to specify ther own tags.

mkurz commented on web/play-filters-helpers/src/main/scala/play/filters/ip/IPFilter.scala:
```diff
@@ -0,0 +1,145 @@
+     * "2001:cdba:0:0:0:0:3257:9652"
+     * "2001:cdba::3257:9652"
+     * You can easily test this in jshell with java.net.InetAddress.getByName("<ip>").getAddress();
+     */
```
This is important. Because the same IPv6 addresses can be writtein various ways,  we need to check per byte not per string.

In case we ever need to work on more complicated IPv6 address stuff (like checking for ranges...) it's probably better to just use a library like https://github.com/seancfoley/IPAddress

mkurz commented on web/play-filters-helpers/src/main/scala/play/filters/ip/IPFilter.scala:
```diff
@@ -0,0 +1,145 @@
+    val whiteList =
+      ipConfig.getOptional[Seq[String]]("whiteList").getOrElse(Seq.empty).map(InetAddress.getByName(_).getAddress())
+    val blackList =
+      ipConfig.getOptional[Seq[String]]("blackList").getOrElse(Seq.empty).map(InetAddress.getByName(_).getAddress())
```
By using `InetAddress.getByName(_).getAddress()` the white and blacklist will be a array of bytes representing a IP address, which we can [easily compare](https://stackoverflow.com/a/1790040/810109) to the remote ip address.

There is one thing that should be mentioned, however not relevant for us:
When the String you pass to `InetAddress.getByName()` is not an IP address, a DNS lookup would occur, however, according to the Java docs "_If a literal IP address is supplied, only the validity of the address format is checked._", which is the case for our white and blacklist, so we are ok with that. Even if it would do a dns lookup, this list should be build eagerly when the class gets initalized once, so even that wouldn't be too bad. However, I also checked locally with jshell and can confirm it will NOT do any dns lookups when just passing IP addresses.
Also see https://stackoverflow.com/a/52135243/810109 and https://stackoverflow.com/a/60397072/810109

mkurz commented on web/play-filters-helpers/src/main/scala/play/filters/ip/IPFilter.scala:
```diff
@@ -0,0 +1,145 @@
+        }
+      } else {
+        // The whitelist is defined. We accept the IP if there is a matching whitelist entry.
+        whiteList.exists(JArrays.equals(_, req.connection.remoteAddress.getAddress))
```
Even when using Scala, Java's `Arrays.equals(...)` is the fastes way to compar arrays: https://stackoverflow.com/q/5393243/810109

mkurz commented on web/play-filters-helpers/src/test/scala/play/filters/ip/IPFilterSpec.scala:
```diff
@@ -0,0 +1,501 @@
+                 |play.filters.ip.whiteList = [ ]
+                 |play.filters.ip.blackList = [ "192.168.0.3" ]
+                 |play.filters.ip.routeModifiers.whiteList = [ ]
+                 |play.filters.ip.routeModifiers.blackList = [ "checkip" ]
```
Added various tests for the `routeModifiers` white and blacklist ad well.

mkurz commented on web/play-filters-helpers/src/test/scala/play/filters/ip/IPFilterSpec.scala:
```diff
@@ -0,0 +1,501 @@
+                 |play.filters.ip.whiteList = [ "2001:cdba::3257:9652" ]
+      """.stripMargin)
+    ) with Injecting {
+      val req: FakeRequest[AnyContentAsEmpty.type] = request("/", "2001:cdba:0:0:0:0:3257:9652")
```
Starting form here, added various tests to see if the filter correctly compares IPv6 addresses which a written in different notations (padding zeros, skipping parts, etc.)
